Added builder for bpf probe

nathan-b opened the pull request.

gianlucaborello commented on driver/bpf/Dockerfile:
```diff
@@ -0,0 +1,46 @@
+FROM debian:unstable
```
I believe the path chosen for all these files (`driver/bpf/`) might not be the best one, I see no point in setting the build context of Docker to this directory, it makes you think the Docker build will somehow need the other source files at this location, which it doesn't. The best path for this would probably be one of `ROOT/docker/ebpf-probe-builder`, `ROOT/test/ebpf-probe-builder`, `ROOT/scripts/ebpf-probe-builder`. Further, these scripts seem to be hardcoded to work with the agent (e.g. `ENV AGENT_VERSION 0.1.1dev` here and `/opt/draios/src/draios-agent-0.1.1dev` later), so perhaps they should be moved in the agent repository or properly refactored to make sure they can be used against sysdig as well.

gianlucaborello commented on driver/bpf/Dockerfile:
```diff
@@ -0,0 +1,46 @@
+RUN apt-get update \
+ && apt-get dist-upgrade -y \
+ && apt-get install -y --no-install-recommends \
+    apt-utils \
```
Minor formatting issue: these packages are indented with spaces, the rest is indented with tabs.

gianlucaborello commented on driver/bpf/Dockerfile:
```diff
@@ -0,0 +1,46 @@
+ && apt-get dist-upgrade -y \
+ && apt-get install -y --no-install-recommends \
+    apt-utils \
+    cmake \
```
What is the logic behind tweaking this dependency list? It's certainly ok to do so, since this container will never be distributed to a customer, but I just wanted to understand the rationale.

In particular, it seems like you added some packages that are not needed to compile the bpf probe (e.g. `cmake`, `g++`), removed some packages that are not needed to compile the bpf probe (e.g. `dkms`) and kept some other packages that are also not needed to compile the bpf probe (e.g. everything else besides `make`, `gcc`, `llvm` and `clang`). Is it because you envision this container with the ability of doing more than just building the probe or something else?

nathan-b replied:
My original thought was to have this container build both the kmod and the probe. I decided to simplify, but didn't remove all the extra dependencies. I'll remove them.

gianlucaborello commented on driver/bpf/build_bpf_probe.sh:
```diff
@@ -0,0 +1,8 @@
+mkdir -p ${HOME}/.sysdig
+
+docker build -t ebpf-probe-builder:latest --pull .
+docker run --rm -it -v ${HOME}/.sysdig:/out -v /opt/draios/src/draios-agent-0.1.1dev:/driver -v /draios:/draios -v /:/host -e BPF_PROBE_FILENAME=bpf_probe.o ebpf-probe-builder:latest
```
Is `/draios` used at all?

gianlucaborello replied:
Can `/host` be set `:ro`?

gianlucaborello commented on driver/bpf/probe-builder-entrypoint.sh:
```diff
@@ -0,0 +1,39 @@
+# Simple script to build the BPF probe. Assumes that all the dependencies
+# and requirements are already satisfied (as they are in the accompanying
+# docker container)
+#
```
Since this is an internal script, can you perhaps consider using `set -euo pipefail` or even `set -exuo pipefail`? Typically we do this for every script that's mean to be consumed internally or via tests, otherwise you get confusing output such as:

```
* Building probe bpf_probe.o
/probe-builder-entrypoint.sh: line 34: cd: /driver/bpf: No such file or directory
Building bpf
make: *** No targets specified and no makefile found.  Stop.
** Done building probe
cp: cannot stat 'probe.o': No such file or directory
Probe is in /home/gianluca/.sysdig/
```

gianlucaborello commented on driver/bpf/probe-builder-entrypoint.sh:
```diff
@@ -0,0 +1,39 @@
+
+for i in $(ls $SYSDIG_HOST_ROOT/usr/src)
+do
+	ln -s $SYSDIG_HOST_ROOT/usr/src/$i /usr/src/$i
```
I'm not a huge fan of dealing with `/usr/src` in this container considering its purpose (rationale in the other comment).

gianlucaborello commented:
Left a few comments, the general idea makes sense.

If I were to do it I'd probably make it a bit more generic. In particular, I wouldn't rely on using any of the `SYSDIG_HOST_ROOT` tricks we use for the runtime container, so I'd probably get rid of the links to `/lib/modules` and `/usr/src` and just mount the kernel source directory directly into the container at a specified path (e.g. /kernel) and then use `KERNELDIR` to point the kernel makefile to that tree, something like:

```
docker run -v /lib/modules/$(uname -r)/build:/kernel -v driver_src:/driver bpf-builder
```

This way, such container becomes more versatile (and the diff likely smaller) and can not only be used against the running kernel on the machine where the tests are being executed, but also against an arbitrary kernel source directory. This makes it very useful for the precompiled probe builder because that's exactly what we do already for the kmod, so we can hook this container in the precompiled script and quickly get precompiled bpf support.

Since you seem to be trying to solve the other specific problem, don't consider this an item that must be fixed, but I am 100% sure eventually (and soon) this container will also have to be used for that use case, so the changes I'm mentioning will likely have to be done.

nathan-b commented:
Comments sound good. Will address them soon.

nathan-b commented:
I've addressed most of the comments. I still have the agent path hard-coded in the docker run string. I would love to fix this. What I have right now is sufficient to get me up and running on Jenkins, but I would be happy to make it more generic if you have any suggestions regarding how I could fix this to be more generic and not agent-dev-build specific.

gianlucaborello commented on docker/ebpf-probe-builder/Dockerfile:
```diff
@@ -0,0 +1,33 @@
+
+MAINTAINER Sysdig <support@sysdig.com>
+
+ENV SYSDIG_HOST_ROOT /host
```
I believe we should be able to get rid of both `SYSDIG_HOST_ROOT` and `HOME`?

gianlucaborello commented on docker/ebpf-probe-builder/Dockerfile:
```diff
@@ -0,0 +1,33 @@
+ && ln -s /usr/bin/clang-7 /usr/bin/clang \
+ && ln -s /usr/bin/llc-7 /usr/bin/llc
+
+RUN rm -df /lib/modules \
```
Same with this, it shouldn't be needed anymore inside this image, unless I'm missing something.

gianlucaborello commented on docker/ebpf-probe-builder/probe-builder-entrypoint.sh:
```diff
@@ -0,0 +1,36 @@
+
+set -euo pipefail
+
+SYSDIG_HOST_ROOT=/host
```
Same here about `SYSDIG_HOST_ROOT`, unless I'm missing something it's not needed anymore.

gianlucaborello commented on docker/ebpf-probe-builder/build_bpf_probe.sh:
```diff
@@ -0,0 +1,8 @@
+mkdir -p ${HOME}/.sysdig
+
+docker build -t ebpf-probe-builder:latest --pull .
+docker run --rm -it -v ${HOME}/.sysdig:/out -v /opt/draios/src/draios-agent-0.1.1dev:/driver -v /lib/modules/$(uname -r)/build:/kernel -v /:/host:ro -e BPF_PROBE_FILENAME=bpf_probe.o ebpf-probe-builder:latest
```
In my opinion, a generic way would be to just take `/opt/draios/src/draios-agent-0.1.1dev` and `/lib/modules/$(uname -r)/build` as input parameters of the script (e.g. $1, $2) instead of leaving them hardcoded. For your tests, you'll call it using these two arguments, whereas for the probe builder we'll call it using the kernel path we're processing at the moment, and the sysdig driver's working directory right after the cmake bootstrap.

gianlucaborello replied:
Also, do we still need the mount against `/`?

gianlucaborello commented:
Thanks for the changes 👍

gianlucaborello commented on docker/ebpf-probe-builder/build_bpf_probe.sh:
```diff
@@ -0,0 +1,42 @@
+
+# Options parsing
+while [ -n "${1-}" ]; do
+	case $1 in
```
Minor: this file seems to be using both tabs and spaces for indentation.

nathan-b replied:
I used tabs for indentation and spaces for formatting. Anything indented should have a leading tab.

gianlucaborello commented:
LGTM
